bt_crypto: provide a PSA APIs implementation alternative to TinyCrypt

valeriosetti opened the pull request:
This PR adds a new kconfig symbol named `CONFIG_BT_USE_PSA_API` to allow the end user to specifically request PSA API usage over TinyCrypt for crypto operations in bluetooth. This might be useful in platforms/scenarios which already have some PSA API provider (ex: TF-M or Mbed TLS + `CRYPTO_C`) in order to reduce memory footprint and not duplicate crypto libraries for no reason.

For sake of example here's a memory footprint comparison on some BT test cases on the `nrf9160dk/ns` platform (which has TFM):
```
west build -p always -b nrf9160dk/nrf9160/ns -T tests/bluetooth/bt_crypto/bluetooth.bt_crypto
...
Memory region         Used Size  Region Size  %age Used
           FLASH:       44996 B       192 KB     22.89%
             RAM:        8672 B       128 KB      6.62%
        IDT_LIST:          0 GB        32 KB      0.00%
```
```
west build -p always -b nrf9160dk/nrf9160/ns -T tests/bluetooth/bt_crypto/bluetooth.bt_crypto.psa
...
Memory region         Used Size  Region Size  %age Used
           FLASH:       44968 B       192 KB     22.87%
             RAM:        8696 B       128 KB      6.63%
        IDT_LIST:          0 GB        32 KB      0.00%
```
```
west build -p always -b nrf9160dk/nrf9160/ns -T tests/bluetooth/gatt/bluetooth.gatt
...
Memory region         Used Size  Region Size  %age Used
           FLASH:       81196 B       192 KB     41.30%
             RAM:       13436 B       128 KB     10.25%
        IDT_LIST:          0 GB        32 KB      0.00%
```
```
$ west build -p always -b nrf9160dk/nrf9160/ns -T tests/bluetooth/gatt/bluetooth.gatt.psa
...
Memory region         Used Size  Region Size  %age Used
           FLASH:       79132 B       192 KB     40.25%
             RAM:       13156 B       128 KB     10.04%
        IDT_LIST:          0 GB        32 KB      0.00%
```
```
west build -p always -b nrf5340dk/nrf5340/cpuapp/ns -T tests/bluetooth/mesh/basic/bluetooth.mesh.gatt
...
Memory region         Used Size  Region Size  %age Used
           FLASH:      179692 B       192 KB     91.40%
             RAM:       35758 B       192 KB     18.19%
        IDT_LIST:          0 GB        32 KB      0.00%
```
```
west build -p always -b nrf5340dk/nrf5340/cpuapp/ns -T tests/bluetooth/mesh/basic/bluetooth.mesh.gatt.psa
...
Memory region         Used Size  Region Size  %age Used
           FLASH:      171932 B       192 KB     87.45%
             RAM:       35446 B       192 KB     18.03%
        IDT_LIST:          0 GB        32 KB      0.00%
```

valeriosetti committed:
bt-crypto: add option to use PSA APIs instead of TinyCrypt

This commit adds CONFIG_BT_USE_PSA_API to allow the end
user to prefer PSA APIs over TinyCrypt for crypto operations
in bluetooth. Of course, this is possible only if
a PSA provider is available on the system, i.e.
CONFIG_PSA_CRYPTO_CLIENT is set.

This commit also extends tests/bluetooth/bt_crypto adding
a test case for PSA.

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

tomi-font commented on subsys/bluetooth/crypto/Kconfig:
```diff
@@ -2,9 +2,23 @@
+	help
+	  This option enables the Bluetooth Cryptographic Toolbox.
+
+config BT_CRYPTO_PSA
```
Thinking out loud. How about putting `BT_CRYPTO_PSA` and `BT_CRYPTO_TINYCRYPT` under a choice, that itself could be guarded by an `if BT_CRYPTO`?

valeriosetti replied:
> There being a prompt or not doesn't change whether the user needs to manually select anything.

Yes, but if the user wants to take advantage of Mbed TLS it's enough to enable `CONFIG_MBEDTLS` in his/her build (I guess that most of the times Mbed TLS is used also for other purposes, so it's just there ready to be used).
`CONFIG_BT_CRYPTO` was not user-selectable before and it is not now; so do the new options that I added for the crypto backend: they are enabled based on what crypto provider is available in the system. This seems to be the most natural continuation of what was available before. Wdyt?

> Removing the prompt just prevents (to some extent) the user from assigning the Kconfig option.

Actually removing the prompt does not allow the user to select something even in `conf` file.

> And regarding the prompt, I think that you could still keep the particular choice options promptless

I tried, but that seems not possible.

tomi-font replied:
> CONFIG_BT_CRYPTO was not user-selectable before and it is not now; so do the new options that I added for the crypto backend: they are enabled based on what crypto provider is available in the system. This seems to be the most natural continuation of what was available before. Wdyt?

Yeah, I think it's good this way.

> Actually removing the prompt does not allow the user to select something even in conf file.

Here do you mean in `*.conf` files? If so that's the expected effect of promptless symbols. :slightly_smiling_face: 


Regarding the `choice` option (which is really the only thing I am interested about here). Is it so that you cannot make a choice promptless, which is the reason why you want to stick to individual options?

valeriosetti replied:
> Regarding the choice option (which is really the only thing I am interested about here). Is it so that you cannot make a choice promptless, which is the reason why you want to stick to individual options?

Yes, it seems that it's not possible to make the choice promptless. I cannot complain though as I agree on the fact that `choice` means/requires user interaction (or at least the possibility to decide), which is what I'm not providing here.

tomi-font replied:
Okay. Having now tested things a bit locally, I get why you did what you did. This keeps the use of `BT_CRYPTO` identical to before. Less breaking changes can be good.

tomi-font commented on subsys/bluetooth/crypto/Kconfig:
```diff
@@ -2,9 +2,23 @@
 # SPDX-License-Identifier: Apache-2.0
 
 config BT_CRYPTO
+	bool "Select crypto backend for Bluetooth Cryptographic Toolbox"
```
`BT_CRYPTO` used to be promptless, is there a reason to give it a prompt?

valeriosetti replied:
Right, this is a leftover from a change I tried during the development. I will revert it

tomi-font commented on tests/bluetooth/bt_crypto/bt-crypto-mbedtls.conf:
```diff
@@ -0,0 +1,10 @@
+CONFIG_MBEDTLS=y
+CONFIG_MBEDTLS_PSA_CRYPTO_C=y
+CONFIG_MBEDTLS_PSA_CRYPTO_EXTERNAL_RNG=y
```
Is this one mandatory?

frkv replied:
RNG is mandatory to support for bluetooth pairing and connection (link) establishment and probably a lot of other use-cases.

tomi-font replied:
Yes, but my question is why `EXTERNAL` and not `LEGACY`?

valeriosetti replied:
The idea was to keep the PSA crypto implementation to a minimum. Using legacy means add few other modules to Mbed TLS like entropy and ctr/hmac-drbg. I can give this a try though.

tomi-font replied:
Yeah, though by using `EXTERNAL` you add the dependency on `CONFIG_CSPRNG_ENABLED` in the test, so I don't know whether it will actually ever be run in CI? From that point of view, it could be nice if the test could run on `native_sim`.

tomi-font replied:
And here it's a test so I don't think we care much about code size as long as it doesn't overflow... Unless you want to measure the difference between the TC and PSA variants using the unmodified test, obviously.

valeriosetti replied:
> hough by using `EXTERNAL` you add the dependency on `CONFIG_CSPRNG_ENABLED` in the test, so I don't know whether it will actually ever be run in CI?

Just for the record, testing just the mbedtls case on _all_ platform with `twister -c -s tests/bluetooth/bt_crypto/bluetooth.bt_crypto.mbedtls --all` makes the test built and _executed_ on 4 of them:
```
INFO    - 4 of 733 test configurations passed (100.00%), 0 failed, 0 errored, 729 skipped with 0 warnings in 12.16 seconds
[...]
INFO    - 4 test configurations executed on platforms, 0 test configurations were only built.
```

but I can try legacy RNG to see if this way I can make it tested on more platforms.

valeriosetti replied:
I changed to legacy RNG in https://github.com/zephyrproject-rtos/zephyr/pull/73378/commits/4ff681715c66aeb93caaf411d8415c1bc2c21c0f. However I still had to add `CONFIG_ENTROPY_HAS_DRIVER` filter on Mbed TLS because this is needed to use some valid random generator (ex: `ENTROPY_DEVICE_RANDOM_GENERATOR` or `XOSHIRO_RANDOM_GENERATOR` and not `TIMER_RANDOM_GENERATOR`).
In this way the mbedtls test is still run on 4 platforms out of 733, so it seems that this change did not extend "testing coverage" to include more platforms. Let me know if you prefer this new option or the previous one, please.

tomi-font replied:
Okay, thanks for testing out. I don't have a preference.

valeriosetti committed:
doc: update migration guide about changes to bt-crypto

Update migration guide about the support to PSA functions
introduced in bt-crypto.

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

tomi-font commented on tests/bluetooth/bt_crypto/bt-crypto-mbedtls.conf:
```diff
@@ -0,0 +1,10 @@
+CONFIG_ENTROPY_GENERATOR=y
+CONFIG_MBEDTLS_ZEPHYR_ENTROPY=y
+
+CONFIG_MBEDTLS_TLS_VERSION_1_2=n
+CONFIG_MBEDTLS_KEY_EXCHANGE_RSA_ENABLED=n
+CONFIG_MBEDTLS_CIPHER_DES_ENABLED=n
+CONFIG_MBEDTLS_CIPHER_MODE_CBC_ENABLED=n
+CONFIG_MBEDTLS_PK_WRITE_C=n
```
Make sure to remove those once https://github.com/zephyrproject-rtos/zephyr/pull/72475 is merged.

valeriosetti replied:
Since #72475 was merged, I just rebased the PR removing these lines

valeriosetti commented:
CI is currently failing in BabbleSim tests due to https://github.com/zephyrproject-rtos/zephyr/issues/73603

valeriosetti committed:
trusted-firmware-m: select PSA_CRYPTO_CLIENT when BUILD_WITH_TFM

TF-M is a PSA API provider alternative to Mbed TLS one. As
a consequence when CONFIG_BUILD_WITH_TFM is set
CONFIG_PSA_CRYPTO_CLIENT should be set as well.

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

valeriosetti committed:
bt-host: add option to use PSA APIs instead of TinyCrypt

By enabling CONFIG_BT_USE_PSA_API the user can specify to use
PSA APIs instead of TinyCrypt for crypto operations in bluetooth
host module.

This commit also extends tests/bluetooth/gatt in order to
add a PSA test.

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

de-nordic commented on subsys/bluetooth/crypto/bt_crypto.c:
```diff
@@ -17,7 +21,33 @@
+#if defined(CONFIG_BT_USE_PSA_API)
+int bt_crypto_aes_cmac(const uint8_t *key, const uint8_t *in, size_t len, uint8_t *out)
+{
+	mbedtls_svc_key_id_t key_id;
```
Why `mbedtls_svc_key_id_t` is used instead of `psa_key_id_t`?

valeriosetti replied:
Ops, sorry, I worked in Mbed TLS scope for too long so this came up automatically :D. I will fix this

valeriosetti committed:
bluetooth: remove BT_TINYCRYPT_ECC in bluetooth mesh tests

Some bluetooth test were using BT_TINYCRYPT_ECC without also
setting BT_ECC. This means that BT_TINYCRYPT_ECC gets disabled
as it depends on BT_ECC.
This commit fix this by removing BT_TINYCRYPT_ECC in all test
for bluetooth mesh.

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

valeriosetti committed:
bt: hci_ecc: add option to use PSA APIs instead of TinyCrypt

This commit adds CONFIG_BT_USE_PSA_API to allow the end
user to prefer PSA APIs over TinyCrypt for crypto operations
in bluetooth. Of course, this is possible only if
a PSA provider is available on the system, i.e.
CONFIG_PSA_CRYPTO_CLIENT is set.

This commit also extends
tests/bluetooth/mesh/basic/bluetooth.mesh.gatt adding a specific
case using PSA.

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

alxelax reviewed (changes requested):
I see dependency mesh on Hosts's ECC has been moved back. This PR adds the issue that is described here: https://github.com/zephyrproject-rtos/zephyr/issues/23292
The issue was fixed by having independent ECC in mesh and host. This PR adds dependency back.
Also, mbedTLS PSA doesn't support ITS based on Zephyr's settings or similar functionality (actually it supports only posix files). 
No sense to use any PSA from Host in mesh since mesh devices should be able to restore its state after power off\on cycle. 
In case of the suggested crypto usage (even based on PSA) BLE Mesh will have to store raw key values in none secure domain (even secure domain exists). Usage of PSA doesn't have sense, it adds only footprint with absolutely the same crypto level as Tinycrypt.
There was added its own support of PSA in Mesh some time ago: https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/bluetooth/mesh/crypto_psa.c
I'm quite sure this implementation should be used until Host's crypto can suggest ability to manipulate key attributes and mbedTLS will support proper ITS implementation.

valeriosetti commented:
@alxelax thanks for the review!
I'm not an expert at all of bluetooth (I mostly worked on Mbed TLS & crypto so far), so please apologize if I say something uncorrect in the following.

> mbedTLS PSA doesn't support ITS based on Zephyr's settings or similar functionality (actually it supports only posix files).

Yes, I know, but my purpose here was to use PSA only on platforms which have TFM, not to provide PSA through MbedTLS. That's why I made `CONFIG_BT_USE_PSA_API` optional and not enabled by any other symbol.

> Usage of PSA doesn't have sense, it adds only footprint with absolutely the same crypto level as Tinycrypt.

The example I added in the description for `nrf5340dk/nrf5340/cpuapp/ns` on `tests/bluetooth/mesh/basic/bluetooth.mesh.gatt` seems to suggest the opposite actually. Am I missing something?

This work was mostly intended to provide a PSA alternative to TinyCrypt since it is going to be deprecated sooner or later (#43712 #73241). The replacement of TinyCrypt with PSA that I have made is surely pretty basic as I replaced one call with the other, but for the same reason I expect it to work the same for what concerns key storage. 
Is this completely unacceptable?

frkv commented:
> I see dependency mesh on Hosts's ECC has been moved back. This PR adds the issue that is described here: #23292 The issue was fixed by having independent ECC in mesh and host. This PR adds dependency back. Also, mbedTLS PSA doesn't support ITS based on Zephyr's settings or similar functionality (actually it supports only posix files). No sense to use any PSA from Host in mesh since mesh devices should be able to restore its state after power off\on cycle. In case of the suggested crypto usage (even based on PSA) BLE Mesh will have to store raw key values in none secure domain (even secure domain exists). Usage of PSA doesn't have sense, it adds only footprint with absolutely the same crypto level as Tinycrypt. There was added its own support of PSA in Mesh some time ago: https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/bluetooth/mesh/crypto_psa.c I'm quite sure this implementation should be used until Host's crypto can suggest ability to manipulate key attributes and mbedTLS will support proper ITS implementation.

_Short version: We will add PSA secure storage APIs to get ITS support for all devices (that doesn't use TF-M)_

**Long version:**
PSA crypto is decided to be the defacto crypto API in Zephyr. TinyCrypt is unsupported and will be removed. Any shortcomings with regards to improper handling of sensitive keys (i.e. lack of isolation) will be addressed in the Bluetooth community or explicitly in the Zephyr implementation of said features (BLE host or mesh).

Storing the key in clear-text in non-secure images is counter to coming regulatory changes in EU, US and Asian markets and is generally a weak security design. Follow up work to this PR is addition of key-storage APIs for non TF-M devices. The expressed intent is for Zephyr to support `PSA secure storage APIs`, with an addition that you can store these in encrypted form at-rest (if your platform supports it). This is tied directly into PSA crypto API support (that does the key management APIs)

In PSA secure storage APIs and in PSA crypto key management APIs, there are options to counteract any shortcomings in the specification and/or implementation with regards to handling of keys (e.g. handling keys without isolation). Multiple components can make use of a single generated key (for `ECDH `and `ECDSA`, although this is generally not recommended, but may still be accepted/allowed (Please see _NIST SP 800-56A Rev. 3
Recommendation for Pair-Wise Key-Establishment Schemes Using Discrete Logarithm Cryptography_). Additionally you have the option to designate any key as `EXPORTABLE` in case the current version of the code expects to have access to the key (e.g. in a non-secure environment, where the protocol lives)

Using a single static private key for multiple ECDH links (in LE secure connection) is even recommended against in the BLE spec. There they argue for the case that these keys should be changed after a certain number of failing attempts.

ECDH by design doesn't require a single static private key. The private key on client/server side can be handled fully ephemerally (erased after use). It is the established shared secret that is the item to store persistently and our aim is to store it so that it is isolated from the non-secure portions of the program on the devices that can support this.

valeriosetti commented:
> Please remove all CONFIG_BT_ECC dependencies in configuration files related to mesh.

I added `CONFIG_BT_ECC` in those tests because I noticed the following:
```
warning: BT_TINYCRYPT_ECC (defined at subsys/bluetooth/host/Kconfig:913) was assigned the value 'y'
but got the value 'n'. Check these unsatisfied dependencies: BT_ECC (=n). See
http://docs.zephyrproject.org/latest/kconfig.html#CONFIG_BT_TINYCRYPT_ECC and/or look up
BT_TINYCRYPT_ECC in the menuconfig/guiconfig interface. The Application Development Primer, Setting
Configuration Values, and Kconfig - Tips and Best Practices sections of the manual might be helpful
too.
```
If I remove `CONFIG_BT_ECC`, can I also remove `BT_TINYCRYPT_ECC` from those tests in order to avoid the warning?

alxelax commented:
>> 
> If I remove `CONFIG_BT_ECC`, can I also remove `CONFIG_BT_TINYCRYPT_ECC` from those tests in order to avoid the warning?

Yep, I guess `CONFIG_BT_TINYCRYPT_ECC` just leftover of previous dependencies. It is not required anymore.

valeriosetti commented:
> > > 
> > 
> > 
> > If I remove `CONFIG_BT_ECC`, can I also remove `CONFIG_BT_TINYCRYPT_ECC` from those tests in order to avoid the warning?
> 
> Yep, I guess `CONFIG_BT_TINYCRYPT_ECC` just leftover of previous dependencies. It is not required anymore.

Ok, good, that's what I did in https://github.com/zephyrproject-rtos/zephyr/pull/73378/commits/7aa055b2cc98e551e73c87737ca4874864a828dd then ;)

alxelax commented on subsys/bluetooth/crypto/CMakeLists.txt:
```diff
@@ -4,6 +4,10 @@ zephyr_library()
 zephyr_library_sources(bt_crypto.c)
 
+if(CONFIG_BT_USE_PSA_API)
+    zephyr_link_libraries(mbedTLS)
```
Why is this required if `CONFIG_BT_USE_PSA_API` is enabled for platform with support of TF-M based PSA?

valeriosetti replied:
This is only because I need `psa/crypto.h` header to be available. I tried to use the TF-M one in another PR, but without much success. I can give it another try though

valeriosetti replied:
I tried replacing `zephyr_link_libraries(mbedTLS)` with
```
zephyr_include_directories(
        $<TARGET_PROPERTY:tfm,TFM_BINARY_DIR>/api_ns/interface/include
    )
```
but the build fails due to the fact that:
```
zephyr/build/tfm/api_ns/interface/include/mbedtls/build_info.h:106:10: fatal error: mbedtls/mbedtls_config.h: No such file or directory
  106 | #include "mbedtls/mbedtls_config.h"
```

Side note: the same happens in `samples/bluetooth/mesh/` if I set `CONFIG_MBEDTLS=n` in `prj.conf`.

alxelax replied:
Shouldn't they be added together as it has been done in mesh:
```
zephyr_library_link_libraries_ifdef(CONFIG_MBEDTLS mbedTLS)

if (CONFIG_BUILD_WITH_TFM)
  target_include_directories(${ZEPHYR_CURRENT_LIBRARY} PRIVATE
    $<TARGET_PROPERTY:tfm,TFM_BINARY_DIR>/api_ns/interface/include
  )
endif()
```
https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/bluetooth/mesh/CMakeLists.txt#L128-L134

When I, as user, choose TF-M I do not configure anything related to mbedTLS (and even do not choose it explicitly) and vice versa.

alxelax commented on subsys/bluetooth/host/CMakeLists.txt:
```diff
@@ -111,6 +111,6 @@ endif()
 # In order to compile Bsim tests with these test features
 # and PSA enabled, the libraries must be linked.
-if(CONFIG_BT_MESH AND CONFIG_BT_TESTING)
+if((CONFIG_BT_MESH AND CONFIG_BT_TESTING) OR CONFIG_BT_USE_PSA_API)
```
The same question here. Why is it required if TF-M is used?
This was added for mesh<->host dependencies for BabbleSim tests. BabbleSim tests are built for POSIX based platforms. The only mbedTLS is supported for them. 
Since your PR adds PSA support in Host for both pure mbedTLS and TF-M I guess condition here should consider the PSA implementation provider.

alxelax commented on subsys/bluetooth/host/crypto.c:
```diff
@@ -111,6 +136,88 @@ int bt_rand(void *buf, size_t len)
+	psa_set_key_type(&attr, PSA_KEY_TYPE_AES);
+	psa_set_key_bits(&attr, 128);
+	psa_set_key_usage_flags(&attr, PSA_KEY_USAGE_ENCRYPT);
+	psa_set_key_algorithm(&attr, PSA_ALG_SHA_256);
```
Why different crypto algorithms are used in Host with PSA (sha256) and without PSA (aes128)?

valeriosetti replied:
Ops, sorry, that's definitely an error, of course. I will fix this

valeriosetti replied:
Question: since I don't have any physical board which supports both bluetooth and TFM, do you think there is a way I can test these changes through emulation? Test cases that I modified in this PR only build tests, which is not super bad, but this is not enough to find errors like the one you just caught with this comment.

`nrf5340dk/nrf5340/cpuapp/ns` should be a good target and since that's also used in `bsim` I thought that this could be a nice way to test. However I didn't find any `bsim` test for the `nrf5340dk` platform which can test the changes I'm introducing with this PR. Am I missing something?

jori-nordic replied:
@aescolar can probably help you with that, he recently refactored nrf53/54 bsim tests.

alxelax replied:
I'm afraid nrf53 support as bsim emulated chip is on ready yet to full extent. It doesn't support TrustZone, does it?
Nevertheless, I'm pretty sure, it is possible to use\run Host's bsim tests with `CONFIG_BT_USE_PSA_API=y`.
If PSA was added correctly over KConfig\Cmake they should be built without problem (excluding ITS that is not thread safe in mbedTLS). We added this ability into mesh bsim tests. I do not think there is big difference between mesh and hosts tests.

valeriosetti replied:
Thanks a lot for the suggestion!! I added a new commit (the last one - https://github.com/zephyrproject-rtos/zephyr/pull/73378/commits/cb6f66d2ec636029928e1ef044fce6553a743ac9) to add a bsim test case for `CONFIG_BT_USE_PSA_API` for the nrf52840dk platform.

alxelax reviewed (commented):
It seems better, but there are still several questions.

valeriosetti committed:
bsim: add tests for BT_USE_PSA_API

Add a couple PSA overlay configs for the BT tests in order to evaluate
PSA API support introduced by CONFIG_BT_USE_PSA_API. These test are
performed on nrf52840dk platform.

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

jhedberg commented on subsys/bluetooth/host/crypto.c:
```diff
@@ -111,6 +136,88 @@ int bt_rand(void *buf, size_t len)
+
+	sys_memcpy_swap(tmp, plaintext, 16);
+
+	status = psa_cipher_encrypt(key_id, PSA_ALG_SHA_256, tmp, 16,
```
Seems like you still didn't fix this, i.e. it's referencing SHA_256 when it should be using AES_128.

valeriosetti replied:
Ops, sorry, you are absolutely right. This clearly means that the test that I added using PSA APIs is not checking this path. I will investigate and add another one.

jhedberg commented on subsys/bluetooth/host/crypto.c:
```diff
@@ -111,6 +136,88 @@ int bt_rand(void *buf, size_t len)
+		return -EINVAL;
+	}
+
+	status = psa_cipher_encrypt(key_id, PSA_ALG_SHA_256, plaintext, 16,
```
Same here.

alxelax reviewed (commented):
LGTM ATM. When you fix the algorithm type that Johan and I pointed, I'll approve PR.

valeriosetti commented:
> LGTM ATM. When you fix the algorithm type that Johan and I pointed, I'll approve PR.

Fixed. I extended the last commit adding a new test also for encryption using PSA APIs and it works on nrf52840dk platform.

carlescufi commented:
@jhedberg PTAL

carlescufi commented:
@alxelax @jori-nordic please review

alxelax commented on tests/bsim/bluetooth/hci_uart/compile.sh:
```diff
@@ -10,9 +10,11 @@ set -ue
 source ${ZEPHYR_BASE}/tests/bsim/compile.source
 
-app=tests/bsim/bluetooth/ll/conn conf_file=prj_split_hci_uart.conf \
+# app=tests/bsim/bluetooth/ll/conn conf_file=prj_split_hci_uart.conf \
```
Please uncomment

frkv commented on subsys/bluetooth/host/CMakeLists.txt:
```diff
@@ -108,6 +108,15 @@ if(CONFIG_BT_CONN_DISABLE_SECURITY)
 
+if(CONFIG_BT_USE_PSA_API)
+    zephyr_library_link_libraries_ifdef(CONFIG_MBEDTLS mbedTLS)
+    if (CONFIG_BUILD_WITH_TFM)
```
NIT: zephyr_library_include_directories_ifdef can be used there instead to some small code cleanup

zephyr_library_include_directories is using PRIVATE as well as ${ZEPHYR_CURRENT_LIBRARY}

jori-nordic commented on subsys/bluetooth/crypto/bt_crypto.c:
```diff
@@ -17,7 +21,36 @@
 LOG_MODULE_REGISTER(bt_crypto);
 
+#if defined(CONFIG_BT_USE_PSA_API)
+int bt_crypto_aes_cmac(const uint8_t *key, const uint8_t *in, size_t len, uint8_t *out)
```
could you instead add two new files with `bt_crypto_aes_cmac` implementations for psa and for tinycrypt?

jori-nordic commented on subsys/bluetooth/host/crypto.c:
```diff
@@ -111,6 +136,102 @@ int bt_rand(void *buf, size_t len)
 }
 #endif /* CONFIG_BT_HOST_CRYPTO_PRNG */
 
+#if defined(CONFIG_BT_USE_PSA_API)
```
can you make a new file `crypto_psa.c` ? You can rename that one to `crypto_tc.c`

jori-nordic commented on subsys/bluetooth/host/gatt.c:
```diff
@@ -754,6 +763,32 @@ static uint8_t gen_hash_m(const struct bt_gatt_attr *attr, uint16_t handle,
 	case BT_UUID_GATT_INCLUDE_VAL:
 	case BT_UUID_GATT_CHRC_VAL:
 	case BT_UUID_GATT_CEP_VAL:
+#if defined(CONFIG_BT_USE_PSA_API)
```
please make a utility function. This applies to the whole PR. I really don't like ifdef soups.

valeriosetti commented:
@frkv @jori-nordic I think I've addressed all your comments. Please take a look ;)

frkv reviewed (approved):
Looks great!

theob-pro commented on subsys/bluetooth/host/gatt.c:
```diff
@@ -693,11 +697,93 @@ static ssize_t cf_write(struct bt_conn *conn, const struct bt_gatt_attr *attr,
 	return len;
 }
 
+#if defined(CONFIG_BT_USE_PSA_API)
```
I think @jori-nordic was saying that the cryptographic specifics should be abstracted away from `gatt.c`.

valeriosetti replied:
I'm not sure. He mentioned a `utility function` in this case, whereas in previous comments he explicitly said to create a separate file. As you (@jori-nordic) prefer, I can fix, this just let me know :)

valeriosetti replied:
Sure, I can work on that!
Just out of my curiosity would this new PR still be in scope of the next release or will it be postponed after the code freeze?

jori-nordic replied:
you should post it right away but it will be merged after, yes.

aescolar commented on tests/bsim/bluetooth/host/gatt/caching/test_scripts/_run_test.sh:
```diff
@@ -19,3 +19,14 @@ Execute ./bs_2G4_phy_v1 -v=${verbosity_level} -s=${simulation_id} \
 
 wait_for_background_jobs
+
+Execute ./bs_${BOARD_TS}_tests_bsim_bluetooth_host_gatt_caching_prj_conf_psa_overlay_conf \
```
I havent read all comments so this may be justified, so please bear with me.
This is used by 6 tests.
Is there a good reason to double those 6? 
Meaning, is there a good reason why this is done in a shared subscript instead of having its own separate test?
(I just want to be sure we do not start increasing CI time a lot without a good reason).

valeriosetti replied:
Well, based on the failures I had when I first introduced PSA tests here, I think that 1 single case for the GATT should be enough. I mean, once I got 1 single GATT test passing, then all the others passed as well. I can improve this part

aescolar replied:
No need for the double psa in the sim_id  (sorry that was just a silly mistake of mine), you don't need to change it back either

aescolar commented on tests/bsim/bluetooth/hci_uart/compile.sh:
```diff
@@ -12,6 +12,8 @@ source ${ZEPHYR_BASE}/tests/bsim/compile.source
 
 app=tests/bsim/bluetooth/ll/conn conf_file=prj_split_hci_uart.conf \
     cmake_extra_args=-DEXTRA_DTC_OVERLAY_FILE=hci-uart.overlay compile
+app=tests/bsim/bluetooth/ll/conn conf_file=prj_split_hci_uart.conf conf_overlay=psa_overlay.conf \
```
Is there any reason to do this in the set of tests meant to test the HCI UART driver? (controller and host in separate nrf52 devices connected over UART)
Or can this be done in the plain ll/conn tests?

valeriosetti replied:
That's the only test that I found in `bsim/bluetooth` that had `CONFIG_BT_HOST_CRYPTO` enabled and this is what I needed to test changes in `subsys/bluetooth/host/crypto_psa.c`

aescolar commented on tests/bsim/bluetooth/host/gatt/caching/test_scripts/gatt_caching_psa_db_hash_read_eatt.sh:
```diff
@@ -0,0 +1,9 @@
+# Copyright 2024 Nordic Semiconductor ASA
+# SPDX-License-Identifier: Apache-2.0
+
+simulation_id="gatt_caching_psa_db_hash_read_eatt" \
```
```suggestion
simulation_id="gatt_caching_psa_db_hash_read_eatt_psa" \
```

valeriosetti commented:
Mmm not sure the failure reported by `bsim` belongs to this PR. I can reproduce it (after I rebase my PR on current `main`) but it seems unrelated to the changes I did here. Am I wrong?

valeriosetti commented:
I added a new commit (https://github.com/zephyrproject-rtos/zephyr/pull/73378/commits/7bf45b7c10bf3027cdc1a204a8ab7416fd1e0be1) because there are still failures in the CI and BabbleSim due to #73571.

ceolin reviewed (dismissed):
\0/

valeriosetti commented:
> I added a new commit ([7bf45b7](https://github.com/zephyrproject-rtos/zephyr/commit/7bf45b7c10bf3027cdc1a204a8ab7416fd1e0be1)) because there are still failures in the CI and BabbleSim due to #73571.

Removing the last commit because it was fixed by 2690cacf659226ee4abf707c0d8618972306657d and now there's a conflict. Therefore I'm also rebasing on `main`

valeriosetti commented:
I think I need a bit of help here: failure on `BabbleSim Tests` seems to be related to the CI not to the code. Am I wrong?

theob-pro commented:
> I think I need a bit of help here: failure on `BabbleSim Tests` seems to be related to the CI not to the code. Am I wrong?

Re-triggered the bsim tests, let's see if it solve the issue

valeriosetti commented:
> > I think I need a bit of help here: failure on `BabbleSim Tests` seems to be related to the CI not to the code. Am I wrong?
> 
> Re-triggered the bsim tests, let's see if it solve the issue

I already did this early this morning, but it failed in the same point. However let's see, maybe this time is better :)

theob-pro commented:
> > > I think I need a bit of help here: failure on `BabbleSim Tests` seems to be related to the CI not to the code. Am I wrong?
> > 
> > 
> > Re-triggered the bsim tests, let's see if it solve the issue
> 
> I already did this early this morning, but it failed in the same point. However let's see, maybe this time is better :)

Oh my bad, I guess it's a problem on main if the problem persist

aescolar commented:
> I think I need a bit of help here: failure on BabbleSim Tests seems to be related to the CI not to the code. Am I wrong?

Sorry that was my bad, a fix is queued in here https://github.com/zephyrproject-rtos/zephyr/pull/74289

aescolar commented:
@valeriosetti please rebase. Currently the old workflow file is still being picked and it will fail again.
